README.md
# badger

A Clojure app for hackathon

## Usage

hum dee dum... todo

## License

Copyright © 2016 Meetup/Ken Shih

## Notes
docker build -t badger-app .
docker run -p 3000:3000 --rm --name badger-running badger-app
IP_LOCAL_APP_CONTAINER=192.168.99.100
IP_LOCAL_APP_CONTAINER=(docker-machine ip default)
docker kill badger-running
curl http://192.168.99.100:3000/hi
docker run -p 3000:80 --rm --name badger-running badger-app

## deployment steps
export BUILD_NUMBER=17
make package publish
## need to update version in here:
kubectl apply -f webserver-dply.yaml
kubectl --namespace badger-ken describe deployments webserver
curl http://104.196.11.217:3000/


{
    "mrkdwn": false,
    "text": "joe has earned \n http://www.gummyworm.net/wp-content/uploads/2015/02/Pinguino-png-129x129.png"
}

curl -X POST --data-urlencode payload='{"channel": "#random", "text": "joe has earned \n http://www.gummyworm.net/wp-content/uploads/2015/02/Pinguino-png-129x129.png"}' https://hooks.slack.com/services/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

http://www.http-kit.org/client.html

 (require '[org.httpkit.client :as http])
  (require '[clojure.data.json :as json])

(http/post "https://hooks.slack.com/services/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"
  {:form-params {:payload (json/write-str {:channel "#random" "text" "test from lein"})}})


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-07-13
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-07-13
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/badger/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/badger/compare/0.1.0...0.1.1


test/badger/core_test.clj
(ns badger.core-test
  (:require [clojure.test :refer :all]
            [badger.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/badger/badgerer.clj
(ns badger.badgerer
  "hands out the badge onto the profile pic"
  (:require [quil.core :as q]))

(defn prefix-path [filename] (str "resources/public/" filename))

(def profile-size 600)

(def badge-size 200)

(def orig-file (ref ""))

(def badg-file (ref ""))

(def dest-file (ref ""))

(def should-write (ref false))

(q/defsketch badged-profile
 :size [profile-size profile-size]
 :setup (fn []
          (q/frame-rate 10))
 :title "saving to file"
 :draw (fn []
         (when (deref should-write)
           (dosync
             (ref-set should-write false)
             (let [profile-pic (q/load-image (deref orig-file))
                   badge-pic (q/load-image (deref badg-file))
                   result-filename (deref dest-file)
                   badge-pos (- profile-size badge-size)]
               (q/image profile-pic 0 0 profile-size profile-size)
               (q/image badge-pic badge-pos badge-pos badge-size badge-size)
               (q/save result-filename))))))

(defn add-badge
  "receives profile pic filename and badge number, and saves the badged pic and return the filename"
  [profile-filename badge-num]
  (let [result-filename (clojure.string/replace profile-filename #"\." (str "_" badge-num "."))]
    (dosync
      (ref-set orig-file (prefix-path profile-filename))
      (ref-set badg-file (prefix-path (str "badge_" badge-num ".png")))
      (ref-set dest-file (prefix-path result-filename))
      (ref-set should-write true))
    (Thread/sleep 1000)
    result-filename))


src/badger/core.clj
(ns badger.core
  (:use ring.middleware.params
        ring.middleware.resource
        ring.middleware.content-type
        ring.util.response
        ring.adapter.jetty
        clojure.string
        badger.badgerer
        )
  (:gen-class))
(require '[org.httpkit.client :as http])
(require '[clojure.data.json :as json])
(import '[org.apache.commons.lang3 RandomStringUtils])

(declare cmd-default http200 post-one bust summary-badge-numbers mk-image-url)

(defn msg-leader-entry [person-name badge-title url] (format "%s was awarded %s\n%s" person-name badge-title (bust url)))
(defn msg-leader-entry-title [person-name num] (format "%s has %s awards!" person-name num))
(defn msg-award [person-name badge-name] (format "@%s receives the badge '%s'" person-name badge-name))
(defn msg-award-err [] "you need to enter a recipient and an award for this to work")

; mapping to badge #
(def summary-badge-numbers {
  "star" 0
  "crown" 1
  "cupp" 13
  "codekiller" 3
  "grace" 4
  "mentor" 5
  "email" 6
  "onboarded" 7
  "anniversary" 8
  "crawl" 9
  "catchall" 2
  "hack" 10
  })
; "cueen" badge-cc
; "codekiller" badge-codekiller
; "grace" badge-grace
; "mentor" badge-mentor
; "email" badge-email
; "onboarded" badge-onboarded
; "anniversary" badge-anniversary
; "crawl" badge-crawl
; "catchall" badge-catchall
; "hack" badge-hd

(def badge-cc { :title "Coverage Cueen"
                :desc "covered so much code, the code could sleep well"
                :ico-url "badge_crown.png"
                })
(def badge-codekiller { :title "Code Killer"
                :desc "Cleaned up the most code in an H"
                :ico-url "badge_codekiller.png"
                })
(def badge-grace { :title "Amazing Grace"
                :desc "For extraordinary contributions in testing"
                :ico-url "badge_grace.png"
              })
(def badge-mentor { :title "MEME Mentor"
                :desc "Taught a class on testing"
                :ico-url "badge_mentor.png"
                })
(def badge-email { :title "Email Destroyer"
                :desc "Answered 100 support emails in a single day. Able to leap tall buildings in a single bound."
                :ico-url "badge_email.png"
                })
; (def badge- { :title "FAFF Attendee"
;                 :desc "RSVP and attend a FAFF"
;                 :ico-url "badge_.png"
;                 })
(def badge-onboarded { :title "On your way!"
                :desc "Complete your on-boarding training"
                :ico-url "badge_onboarded.png"
                })
(def badge-anniversary { :title "Happy Anniversary"
                :desc "Receive this badge on the day of your work anniversary"
                :ico-url "badge_anniversary.png"
                })
(def badge-crawl { :title "Meetup Crawler"
                :desc "Went on a Meetup Crawl"
                :ico-url "badge_crawl.png"
                })
; (def badge- { :title "Propper"
;                 :desc "Received/Given a Prop at Mupdate"
;                 :ico-url "badge_.png"
;                 })
(def badge-catchall { :title "Gotta Catch Em All"
                :desc "Gain at least 5 badges"
                :ico-url "badge_catchemall.png"
                })
(def badge-hd { :title "Hackday Winner"
                :desc "no one can ever take that from ya!"
                :ico-url "badge_meetupswarm.png"
                })

; lookup of badges by key
(def badges {"cueen" badge-cc
  "codekiller" badge-codekiller
  "grace" badge-grace
  "mentor" badge-mentor
  "email" badge-email
  "onboarded" badge-onboarded
  "anniversary" badge-anniversary
  "crawl" badge-crawl
  "catchall" badge-catchall
  "hack" badge-hd
  })

; users and badges
(def users {
  "kristen" [badge-codekiller badge-mentor badge-hd badge-crawl badge-catchall]
  "ken" [badge-codekiller badge-cc]
  "bryanvelzy" [badge-email badge-mentor badge-anniversary]
  "sak" [badge-mentor badge-grace badge-email badge-onboarded]
  })

; users and badges
(def profile-img-by-user {
  "kristen" "kristen.png"
  "ken" "ken.png"
  "bryanvelzy" "bryanvelzy.png"
  "sak" "sak.png"
  })

; hard-coded fake ordering of leaderboard!
(def top-users ["kristen" "sak" "bryanvelzy" "ken"])



(defn attachJson [title footer icon-url server-name]
  (let [  template (str   "{\"fallback\": \"%s\","
            "\"color\": \"#36a64f\", "
            "\"title\": \"%s\","
            "\"footer\": \"%s\","
            "\"footer_icon\": \"%s\"}")
          ]
          (format template title title footer (mk-image-url icon-url server-name))
  ))

; dtod dyn
(defn listJson[server-name]
  (defn mk-tmpl-with-more [m]
    (str
      "{\"response_type\": \"in_channel\","
      "\"text\": \"Here is the list of possible badges\","
      "\"attachments\": [" m "]"
      "}"))
    (mk-tmpl-with-more
     (join ", " (map #(attachJson (:title %) (:desc %) (:ico-url %) server-name) (vals badges)))
     )
  )

;
; Routing and such follows
;
(def default-response
  (str "{\"response_type\": \"in_channel\",\"text\": \"(default unconfigured response)\" }\n"
  ))

(comment
  "token=gIkuvaNzQIHg97ATvDxqgjtO
  team_id=T0001
  team_domain=example
  channel_id=C2147483705
  channel_name=test
  user_id=U2147483697
  user_name=Steve
  command=/weather
  text=94070
  response_url=https://hooks.slack.com/commands/1234/5678"
  )


(defn cmd-list [server-name] (http200 (listJson server-name)))

(defn channel-name [channel] (str "#" channel))

(defn cmd-top [channel server-name]
  (doseq [u top-users]
    (Thread/sleep 200)
    (post-one (channel-name channel) (msg-leader-entry-title u (count (users u))))
    (doseq [b (users u)]
      (Thread/sleep 200) ;hack to get ordering in ui right
      (post-one (channel-name channel)
        (msg-leader-entry u (:title b) (mk-image-url (:ico-url b) server-name))
        )))
  {:status 200})


(defn post-one [channel text]
  (http/post "https://hooks.slack.com/services/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"
   {:form-params {:payload (json/write-str {:channel channel "text" text})}})
   )

; http utils
(defn http200 [body]
  {:status 200
    :headers {"Content-Type" "application/json"}
    :body (str body)
  })
; cache buster
(defn bust [url]
  (str url "?" (RandomStringUtils/randomAlphanumeric 2)))

; format imge url
(defn mk-image-url [file-name server-name]
  (format "http://%s:3000/%s" server-name file-name)
  )
;
; cmd(s) are expected to
; return a ring "response" object
;
(defn cmd-award [channel text]
  (let [
        terms (split text #" ")
        should-be-3 (count terms)
        channel-name (str "#" channel)
        ]
    (if (= 3 should-be-3)
      (let [name (terms 1) badge (terms 2)]
        (post-one channel-name (msg-award name (:title (badges badge)))))
      (post-one channel-name (msg-award-err)))
    {:status 200})) ; suppress feedback in favor of dynamic feedback

(defn cmd-default []
  {:status 200
    :headers {"Content-Type" "application/json"}
    :body default-response
  }
  )

; profile image creation
(defn cmd-create-profile [text server-name]
  (defn msg [base-url user-name summary-badge]
    {:status 200
      :headers {"Content-Type" "application/json"}
      :body (str "{ "
          "\"response_type\": \"in_channel\", \"text\" : \""
          base-url (add-badge (profile-img-by-user user-name) (summary-badge-numbers summary-badge))
          "\" }"
          )})
    (let [
      terms (split text #" ")
      u (terms 1)
      badge (terms 2)
      ]
      (msg (format "http://%s:3000/" server-name) u badge))
    )

(defn handler [request]
  (let [
    params (:params request)
    server-name (:server-name request)
    channel_id (params "channel_id")
    channel_name (params "channel_name")
    user_id (params "user_id")
    user_name (params "user_name")
    command (params "command")
    text (params "text" "defaut")
    response_url (params "response_url")
    first-text ((split text #" ") 0)
    ]
  (case first-text
    "list" (cmd-list server-name)
    "leaderboard" (cmd-top channel_name server-name)
    "award" (cmd-award channel_name text)
    "response-url" (http200 (str "response-url: " response_url "\n"))
    "create-profile" (cmd-create-profile text server-name)
    (cmd-default)
    )))

(defn router [request]
  (let [ uri (:uri request)
      ]
      (handler request)
    ))

(def app
    (-> router (wrap-resource "public") (wrap-content-type) wrap-params))

(defn -main
  "starts web application"
  [& args]
  (println "starting app")
  (run-jetty app {:port 3000})
  )
